diff --git a/org/apache/axis/description/JavaServiceDesc.java b/org/apache/axis/description/JavaServiceDesc.java
index ba25268..dfb5f2f 100644
--- a/org/apache/axis/description/JavaServiceDesc.java
+++ b/org/apache/axis/description/JavaServiceDesc.java
@@ -707,10 +707,10 @@ public class JavaServiceDesc implements ServiceDesc {
                         }
                         if (paramClass == null) {
                             paramClass = getTypeMapping().getClassForQName(param.getTypeQName(),
-                                                                           type);
+                                                                           null);
                         }
 
-                        if (paramClass != null) {
+                        if (paramClass != null && paramClass != Object.class) {
                             // This is a match if the paramClass is somehow
                             // convertable to the "real" parameter type.  If not,
                             // break out of this loop.
