diff --git a/com/ctc/wstx/sw/BufferingXmlWriter.java b/com/ctc/wstx/sw/BufferingXmlWriter.java
index 345db45..794b4ee 100644
--- a/com/ctc/wstx/sw/BufferingXmlWriter.java
+++ b/com/ctc/wstx/sw/BufferingXmlWriter.java
@@ -462,12 +462,8 @@ public final class BufferingXmlWriter
                         ent = "&amp;";
                         break inner_loop;
                     } else if (c == '>') {
-                        // Let's be conservative; and if there's any
-                        // change it might be part of "]]>" quote it
-                        if (inPtr < 2 || text.charAt(inPtr-2) == ']') {
-                            ent = "&gt;";
-                            break inner_loop;
-                        }
+                        ent = "&gt;";
+                        break inner_loop;
                     }
                 } else if (c >= highChar) {
                     break inner_loop;
@@ -512,13 +508,8 @@ public final class BufferingXmlWriter
                             ent = "&amp;";
                             break;
                         } else if (c == '>') {
-                            /* Let's be conservative; and if there's any
-                             * change it might be part of "]]>" quote it
-                             */
-                            if ((offset == start) || cbuf[offset-1] == ']') {
-                                ent = "&gt;";
-                                break;
-                            }
+                            ent = "&gt;";
+                            break;
                         } else if (c < 0x0020) {
                             if (c == '\n' || c == '\t') { // fine as is
                                 ;
@@ -1400,3 +1391,4 @@ public final class BufferingXmlWriter
         mOutputPtr = ptr;
     }
 }
+/* @generated */
\ No newline at end of file
