diff --git a/server/src/main/java/org/jboss/as/server/deployment/module/ServerDependenciesProcessor.java b/server/src/main/java/org/jboss/as/server/deployment/module/ServerDependenciesProcessor.java
index c61b8fa..aefe140 100644
--- a/server/src/main/java/org/jboss/as/server/deployment/module/ServerDependenciesProcessor.java
+++ b/server/src/main/java/org/jboss/as/server/deployment/module/ServerDependenciesProcessor.java
@@ -38,6 +38,7 @@
 public class ServerDependenciesProcessor implements DeploymentUnitProcessor {
 
     private static ModuleIdentifier SUN_JDK = ModuleIdentifier.create("sun.jdk");
+    private static ModuleIdentifier IBM_JDK = ModuleIdentifier.create("ibm.jdk");
     private static ModuleIdentifier JAVAX_API = ModuleIdentifier.create("javax.api");
     private static ModuleIdentifier JBOSS_LOGGING = ModuleIdentifier.create("org.jboss.logging");
     private static ModuleIdentifier JBOSS_VFS = ModuleIdentifier.create("org.jboss.vfs");
@@ -52,6 +53,7 @@ public void deploy(final DeploymentPhaseContext phaseContext) throws DeploymentU
         final ModuleSpecification moduleSpecification = deploymentUnit.getAttachment(Attachments.MODULE_SPECIFICATION);
         final ModuleLoader moduleLoader = Module.getBootModuleLoader();
         moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, SUN_JDK, false, false, true, false));
+        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, IBM_JDK, false, false, true, false));
         moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, JAVAX_API, false, false, false, false));
         moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, JBOSS_LOGGING, false, false, false, false));
         moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, JBOSS_VFS, false, false, false, false));