diff --git a/hudson/tasks/junit/JUnitResultArchiver.java b/hudson/tasks/junit/JUnitResultArchiver.java
index d3804a9..6f6ed40 100644
--- a/hudson/tasks/junit/JUnitResultArchiver.java
+++ b/hudson/tasks/junit/JUnitResultArchiver.java
@@ -122,7 +122,9 @@ public class JUnitResultArchiver extends Recorder {
 			BuildListener listener) throws InterruptedException, IOException {
 		listener.getLogger().println(Messages.JUnitResultArchiver_Recording());
 		TestResultAction action;
-		
+
+		listener.getLogger().println("Liferay patched " + JUnitResultArchiver.class);
+
 		final String testResults = build.getEnvironment(listener).expand(this.testResults);
 
 		try {
@@ -152,9 +154,6 @@ public class JUnitResultArchiver extends Recorder {
 			}
 
 			action.setData(data);
-
-			CHECKPOINT.block(listener, getDescriptor().getDisplayName());
-
 		} catch (AbortException e) {
 			if (build.getResult() == Result.FAILURE)
 				// most likely a build failed before it gets to the test phase.
@@ -171,7 +170,6 @@ public class JUnitResultArchiver extends Recorder {
 		}
 
 		build.addAction(action);
-		CHECKPOINT.report();
 
 		if (action.getResult().getFailCount() > 0)
 			build.setResult(Result.UNSTABLE);
@@ -216,12 +214,6 @@ public class JUnitResultArchiver extends Recorder {
 		return keepLongStdio;
 	}
 
-	/**
-	 * Test result tracks the diff from the previous run, hence the checkpoint.
-	 */
-	private static final CheckPoint CHECKPOINT = new CheckPoint(
-			"JUnit result archiving");
-
 	private static final long serialVersionUID = 1L;
 
     @Extension
@@ -264,3 +256,4 @@ public class JUnitResultArchiver extends Recorder {
 		}
     }
 }
+/* @generated */
\ No newline at end of file
