diff --git a/name/pachler/nio/file/impl/BSDPathWatchService.java b/name/pachler/nio/file/impl/BSDPathWatchService.java
index 678e558..8b51ca4 100644
--- a/name/pachler/nio/file/impl/BSDPathWatchService.java
+++ b/name/pachler/nio/file/impl/BSDPathWatchService.java
@@ -373,7 +373,7 @@ public class BSDPathWatchService extends PathWatchService{
 						boolean eventsAdded;
 
 						// check if watch key has become invalid
-						if( (fflags & NOTE_DELETE)!=0 || (fflags & NOTE_REVOKE)!=0)
+						if((fflags & NOTE_REVOKE)!=0)
 							eventsAdded = cancelImpl(key, true);
 						else{
 							try {
