diff --git a/freemarker/core/TextBlock.java b/freemarker/core/TextBlock.java
index 9353610..fa80a33 100644
--- a/freemarker/core/TextBlock.java
+++ b/freemarker/core/TextBlock.java
@@ -24,14 +24,14 @@ import freemarker.template.utility.StringUtil;
  * A TemplateElement representing a block of plain text.
  */
 public final class TextBlock extends TemplateElement {
-    private static final char[] EMPTY_CHAR_ARRAY = new char[0];
-    static final TextBlock EMPTY_BLOCK = new TextBlock(EMPTY_CHAR_ARRAY, false);
+    private static final String EMPTY_STRING = "";
+    static final TextBlock EMPTY_BLOCK = new TextBlock(EMPTY_STRING, false);
     // We're using char[] instead of String for storing the text block because
     // Writer.write(String) involves copying the String contents to a char[] 
     // using String.getChars(), and then calling Writer.write(char[]). By
     // using Writer.write(char[]) directly, we avoid array copying on each 
     // write. 
-    private char[] text;
+    private String text;
     private final boolean unparsed;
 
     public TextBlock(String text) {
@@ -39,12 +39,8 @@ public final class TextBlock extends TemplateElement {
     }
 
     public TextBlock(String text, boolean unparsed) {
-        this(text.toCharArray(), unparsed);
-    }
-
-    private TextBlock(char[] text, boolean unparsed) {
         this.text = text;
-        this.unparsed = unparsed;
+		this.unparsed = unparsed;
     }
 
     /**
@@ -87,11 +83,11 @@ public final class TextBlock extends TemplateElement {
     }
 
     TemplateElement postParseCleanup(boolean stripWhitespace) {
-        if (text.length == 0) return this;
+        if (text.isEmpty()) return this;
         int openingCharsToStrip = 0, trailingCharsToStrip=0;
         boolean deliberateLeftTrim = deliberateLeftTrim();
         boolean deliberateRightTrim = deliberateRightTrim();
-        if (!stripWhitespace || text.length == 0 ) {
+        if (!stripWhitespace || text.isEmpty() ) {
             return this;
         }
         if (getParentElement().getParentElement() == null && previousSibling() == null) return this;
@@ -104,7 +100,7 @@ public final class TextBlock extends TemplateElement {
         if (openingCharsToStrip == 0 && trailingCharsToStrip == 0) {
             return this;
         }
-        this.text = substring(text, openingCharsToStrip, text.length - trailingCharsToStrip);
+        this.text = text.substring(openingCharsToStrip, text.length() - trailingCharsToStrip);
         if (openingCharsToStrip > 0) {
             this.beginLine++;
             this.beginColumn = 1;
@@ -134,18 +130,18 @@ public final class TextBlock extends TemplateElement {
                     result = true;
                     int lastNewLineIndex = lastNewLineIndex();
                     if (lastNewLineIndex >=0  || beginColumn == 1) {
-                        char[] firstPart = substring(text, 0, lastNewLineIndex + 1);
-                        char[] lastLine = substring(text, 1+lastNewLineIndex); 
-                        if (StringUtil.isTrimmableToEmpty(lastLine)) {
+                        String firstPart = text.substring(0, lastNewLineIndex + 1);
+                        String lastLine = text.substring(1+lastNewLineIndex);
+                        if (isTrimmableToEmpty(lastLine)) {
                             this.text = firstPart;
                             this.endColumn = 0;
                         } else {
                             int i =0;
-                            while (Character.isWhitespace(lastLine[i])) {
+                            while (Character.isWhitespace(lastLine.charAt(i))) {
                                 i++;
                             }
-                            char[] printablePart = substring(lastLine, i);
-                            this.text = concat(firstPart, printablePart);
+                            String printablePart = lastLine.substring(i);
+                            this.text = firstPart.concat(printablePart);
                         }
                     }
                 }
@@ -175,25 +171,25 @@ public final class TextBlock extends TemplateElement {
                     if (firstLineIndex == 0) {
                         return false;
                     }
-                    if (text.length > firstLineIndex 
-                        && text[firstLineIndex-1] == '\r' 
-                        && text[firstLineIndex] == '\n') 
+                    if (text.length() > firstLineIndex
+                        && text.charAt(firstLineIndex-1) == '\r'
+                        && text.charAt(firstLineIndex) == '\n')
                     {
                         firstLineIndex++;
                     }
-                    char[] trailingPart = substring(text, firstLineIndex);
-                    char[] openingPart = substring(text, 0, firstLineIndex);
-                    if (StringUtil.isTrimmableToEmpty(openingPart)) {
+                    String trailingPart = text.substring(firstLineIndex);
+                    String openingPart = text.substring(0, firstLineIndex);
+                    if (isTrimmableToEmpty(openingPart)) {
                         this.text = trailingPart;
                         this.beginLine++;
                         this.beginColumn=1;
                     } else {
-                        int lastNonWS = openingPart.length -1;
-                        while (Character.isWhitespace(text[lastNonWS])) {
+                        int lastNonWS = openingPart.length() -1;
+                        while (Character.isWhitespace(text.charAt(lastNonWS))) {
                             lastNonWS--;
                         }
-                        char[] printablePart = substring(text, 0, lastNonWS+1);
-                        if (StringUtil.isTrimmableToEmpty(trailingPart)) {
+                        String printablePart = text.substring(0, lastNonWS+1);
+                        if (isTrimmableToEmpty(trailingPart)) {
                         // THIS BLOCK IS HEINOUS! THERE MUST BE A BETTER WAY! REVISIT (JR)
                             boolean trimTrailingPart = true;
                             for (TemplateElement te = this.nextTerminalNode(); 
@@ -209,9 +205,9 @@ public final class TextBlock extends TemplateElement {
                                     break;
                                 }
                             }
-                            if (trimTrailingPart) trailingPart = EMPTY_CHAR_ARRAY;
+                            if (trimTrailingPart) trailingPart = EMPTY_STRING;
                         }
-                        this.text = concat(printablePart, trailingPart);
+                        this.text = printablePart.concat(trailingPart);
                     }
                 }
             }
@@ -220,9 +216,8 @@ public final class TextBlock extends TemplateElement {
     }
     
     private int firstNewLineIndex() {
-        char[] text = this.text;
-        for (int i = 0; i < text.length; i++) {
-            char c = text[i];
+        for (int i = 0; i < text.length(); i++) {
+            char c = text.charAt(i);
             if (c == '\r' || c == '\n' ) {
                 return i;
             }
@@ -231,9 +226,8 @@ public final class TextBlock extends TemplateElement {
     }
 
     private int lastNewLineIndex() {
-        char[] text = this.text;
-        for (int i = text.length - 1; i >= 0; i--) {
-            char c = text[i];
+        for (int i = text.length() - 1; i >= 0; i--) {
+            char c = text.charAt(i);
             if (c == '\r' || c == '\n' ) {
                 return i;
             }
@@ -251,12 +245,12 @@ public final class TextBlock extends TemplateElement {
             return 0;
         }
         ++newlineIndex;
-        if (text.length > newlineIndex) {
-            if (newlineIndex >0 && text[newlineIndex-1] == '\r' && text[newlineIndex] == '\n') {
+        if (text.length() > newlineIndex) {
+            if (newlineIndex >0 && text.charAt(newlineIndex-1) == '\r' && text.charAt(newlineIndex) == '\n') {
                 ++newlineIndex;
             }
         }
-        if (!StringUtil.isTrimmableToEmpty(text, 0, newlineIndex)) {
+        if (!isTrimmableToEmpty(text, 0, newlineIndex)) {
             return 0;
         }
         // We look at the preceding elements on the line to see if we should
@@ -282,7 +276,7 @@ public final class TextBlock extends TemplateElement {
         if (lastNewlineIndex == -1 && beginColumn != 1) {
             return 0;
         }
-        if (!StringUtil.isTrimmableToEmpty(text, lastNewlineIndex + 1)) {
+        if (!isTrimmableToEmpty(text, lastNewlineIndex + 1)) {
             return 0;
         }
         // We look at the elements afterward on the same line to see if we should
@@ -296,15 +290,15 @@ public final class TextBlock extends TemplateElement {
                 return 0;
             }
         }
-        return text.length - (lastNewlineIndex + 1);
+        return text.length() - (lastNewlineIndex + 1);
     }
 
     boolean heedsTrailingWhitespace() {
         if (isIgnorable()) {
             return false;
         }
-        for (int i=0; i<text.length; i++) {
-            char c = text[i];
+        for (int i=0; i<text.length(); i++) {
+            char c = text.charAt(i);
             if (c=='\n' || c=='\r') {
                 return false;
             }
@@ -319,8 +313,8 @@ public final class TextBlock extends TemplateElement {
         if (isIgnorable()) {
             return false;
         }
-        for (int i = text.length -1; i>=0; i--) {
-            char c = text[i];
+        for (int i = text.length() -1; i>=0; i--) {
+            char c = text.charAt(i);
             if (c == '\n' || c == '\r') {
                 return false;
             }
@@ -332,10 +326,10 @@ public final class TextBlock extends TemplateElement {
     }
 
     boolean isIgnorable() {
-        if (text == null || text.length == 0) {
+        if (text == null || text.isEmpty()) {
             return true;
         }
-        if (!StringUtil.isTrimmableToEmpty(text)) {
+        if (!isTrimmableToEmpty(text)) {
             return false;
         }
         boolean atTopLevel = (getParentElement().getParentElement() == null);
@@ -355,23 +349,6 @@ public final class TextBlock extends TemplateElement {
                 element instanceof Comment);
     }
 
-    private static char[] substring(char[] c, int from, int to) {
-        char[] c2 = new char[to - from];
-        System.arraycopy(c, from, c2, 0, c2.length);
-        return c2;
-    }
-    
-    private static char[] substring(char[] c, int from) {
-        return substring(c, from, c.length);
-    }
-    
-    private static char[] concat(char[] c1, char[] c2) {
-        char[] c = new char[c1.length + c2.length];
-        System.arraycopy(c1, 0, c, 0, c1.length);
-        System.arraycopy(c2, 0, c, c1.length, c2.length);
-        return c;
-    }
-    
     boolean isOutputCacheable() {
         return true;
     }
@@ -379,5 +356,24 @@ public final class TextBlock extends TemplateElement {
     boolean isNestedBlockRepeater() {
         return false;
     }
-    
+
+	private static boolean isTrimmableToEmpty(String text) {
+        return isTrimmableToEmpty(text, 0, text.length());
+    }
+
+    private static boolean isTrimmableToEmpty(String text, int start) {
+        return isTrimmableToEmpty(text, start, text.length());
+    }
+
+    private static boolean isTrimmableToEmpty(String text, int start, int end) {
+        for (int i = start; i < end; i++) {
+            if (text.charAt(i) > ' ') {
+                return false;
+            }
+        }
+
+		return true;
+	}
+
 }
+/* @generated */
\ No newline at end of file
diff --git a/freemarker/ext/jsp/JspWriterAdapter.java b/freemarker/ext/jsp/JspWriterAdapter.java
index ad53eed..dc62f96 100644
--- a/freemarker/ext/jsp/JspWriterAdapter.java
+++ b/freemarker/ext/jsp/JspWriterAdapter.java
@@ -24,7 +24,7 @@ import javax.servlet.jsp.JspWriter;
 import freemarker.template.utility.SecurityUtilities;
 
 class JspWriterAdapter extends JspWriter {
-    static final char[] NEWLINE = SecurityUtilities.getSystemProperty("line.separator", "\n").toCharArray();
+    static final String NEWLINE = SecurityUtilities.getSystemProperty("line.separator", "\n");
     
     private final Writer out;
     
@@ -174,4 +174,9 @@ class JspWriterAdapter extends JspWriter {
     {
         out.write(arg0, arg1, arg2);
     }
-}
+
+	public void write(String str, int off, int len) throws IOException {
+		out.write(str, off, len);
+	}
+}
+/* @generated */
\ No newline at end of file
