Index: web/web-core/src/main/java/org/apache/catalina/core/ApplicationDispatcher.java
===================================================================
--- web/web-core/src/main/java/org/apache/catalina/core/ApplicationDispatcher.java	(revision 54151)
+++ web/web-core/src/main/java/org/apache/catalina/core/ApplicationDispatcher.java	(working copy)
@@ -1033,8 +1033,26 @@
 
         // Compute a crossContext flag
         HttpServletRequest hcurrent = (HttpServletRequest) current;
-        boolean crossContext =
-            !(context.getPath().equals(hcurrent.getContextPath()));
+        // -------------------------------------------------------------------------------------------------
+        // boolean crossContext =
+        //     !(context.getPath().equals(hcurrent.getContextPath()));
+
+            boolean crossContext = false;
+            if ((state.outerRequest instanceof ApplicationHttpRequest) ||
+                (state.outerRequest instanceof Request) ||
+                (state.outerRequest instanceof HttpServletRequest)) {
+                HttpServletRequest houterRequest = 
+                    (HttpServletRequest) state.outerRequest;
+                Object contextPath = houterRequest.getAttribute(
+                        RequestDispatcher.INCLUDE_CONTEXT_PATH);
+                if (contextPath == null) {
+                    // Forward
+                    contextPath = houterRequest.getContextPath();
+                }
+                crossContext = !(context.getPath().equals(contextPath));
+            }
+
+        // -------------------------------------------------------------------------------------------------
         //START OF 6364900
         crossContextFlag = Boolean.valueOf(crossContext);
         //END OF 6364900
