diff --git a/org/hibernate/cache/UpdateTimestampsCache.java b/org/hibernate/cache/UpdateTimestampsCache.java
index 31fd175..e4d3b73 100644
--- a/org/hibernate/cache/UpdateTimestampsCache.java
+++ b/org/hibernate/cache/UpdateTimestampsCache.java
@@ -60,6 +60,9 @@ public class UpdateTimestampsCache {
 
 	@SuppressWarnings({"UnnecessaryBoxing"})
 	public void preinvalidate(Serializable[] spaces) throws CacheException {
+		if (spaces.length == 0) {
+			return;
+		}
 		// TODO: to handle concurrent writes correctly, this should return a Lock to the client
 
 		readWriteLock.writeLock().lock();
@@ -83,6 +86,9 @@ public class UpdateTimestampsCache {
 
 	 @SuppressWarnings({"UnnecessaryBoxing"})
 	 public void invalidate(Serializable[] spaces) throws CacheException {
+		if (spaces.length == 0) {
+			return;
+		}
 		//TODO: to handle concurrent writes correctly, the client should pass in a Lock
 
 		 readWriteLock.writeLock().lock();
@@ -106,6 +112,9 @@ public class UpdateTimestampsCache {
 
 	@SuppressWarnings({"unchecked", "UnnecessaryUnboxing"})
 	public boolean isUpToDate(Set spaces, Long timestamp) throws HibernateException {
+		if (spaces.isEmpty()) {
+			return true;
+		}
 		readWriteLock.readLock().lock();
 
 		try {
@@ -155,3 +164,4 @@ public class UpdateTimestampsCache {
 	}
 
 }
+/* @generated */
\ No newline at end of file
