diff --git a/org/apache/jasper/runtime/TldScanner.java b/org/apache/jasper/runtime/TldScanner.java
index 0d3e9c8..c7d6aa0 100644
--- a/org/apache/jasper/runtime/TldScanner.java
+++ b/org/apache/jasper/runtime/TldScanner.java
@@ -343,7 +343,9 @@ public class TldScanner implements ServletContainerInitializer {
      */
     private void scanTlds() throws JasperException {
 
-        mappings = new HashMap<String, String[]>();
+        if (mappings == null) {
+            mappings = new HashMap<String, String[]>();
+        }
 
         // Make a local copy of the system jar cache 
         jarTldCacheLocal.putAll(jarTldCache);
@@ -823,3 +825,4 @@ public class TldScanner implements ServletContainerInitializer {
         }
     }
 }
+/* @generated */
