diff --git a/org/apache/jasper/runtime/JspWriterImpl.java b/org/apache/jasper/runtime/JspWriterImpl.java
index ead4f5e..9d09f67 100644
--- a/org/apache/jasper/runtime/JspWriterImpl.java
+++ b/org/apache/jasper/runtime/JspWriterImpl.java
@@ -139,13 +139,16 @@ public class JspWriterImpl extends JspWriter {
             throw new IllegalArgumentException("Buffer size <= 0");
 	this.response = response;
         allocateCharBuffer();
-        // START OF IASRI 4641975/6172992
-        try {
-            response.setBufferSize(sz);
-        } catch (IllegalStateException ise) {
-            // ignore
-        }
-        // END OF IASRI 4641975/6172992
+
+		if (!response.isCommitted()) {
+			// START OF IASRI 4641975/6172992
+			try {
+				response.setBufferSize(sz);
+			} catch (IllegalStateException ise) {
+				// ignore
+			}
+			// END OF IASRI 4641975/6172992
+		}
     }
 
     void init( ServletResponse response, int sz, boolean autoFlush ) {
@@ -153,13 +156,16 @@ public class JspWriterImpl extends JspWriter {
 	this.autoFlush=autoFlush;
 	this.bufferSize=sz;
         allocateCharBuffer();
-        // START OF IASRI 4641975/6172992
-        try {
-            response.setBufferSize(sz);
-        } catch (IllegalStateException ise) {
-            // ignore
-        }
-        // END OF IASRI 4641975/6172992
+
+		if (!response.isCommitted()) {
+			// START OF IASRI 4641975/6172992
+			try {
+				response.setBufferSize(sz);
+			} catch (IllegalStateException ise) {
+				// ignore
+			}
+			// END OF IASRI 4641975/6172992
+		}
     }
 
     /** Package-level access
@@ -800,3 +806,4 @@ public class JspWriterImpl extends JspWriter {
         }
     }
 }
+/* @generated */
\ No newline at end of file
